bundler/helpers/lib/functions/version_resolver.rb

```diff
@@ -0,0 +1,206 @@
+module Functions
+  class VersionResolver
+    GEM_NOT_FOUND_ERROR_REGEX = /locked to (?<name>[^\s]+) \(/.freeze
+
+    attr_reader :dependency_name, :dependency_requirements,
+                :gemfile_name, :lockfile_name,
+                :dir, :credentials
+
+    def initialize(dependency_name:, dependency_requirements:,
+                   gemfile_name:, lockfile_name:, using_bundler_2:,
+                   dir:, credentials:)
+      @dependency_name = dependency_name
+      @dependency_requirements = dependency_requirements
+      @gemfile_name = gemfile_name
+      @lockfile_name = lockfile_name
+      @using_bundler_2 = using_bundler_2
+      @dir = dir
+      @credentials = @credentials
+    end
+
+    def version_details
+      setup_bundler
+
+      dep = dependency_from_definition
+
+      # TODO: Rewrite this
+      #
+      # If the dependency wasn't found in the definition, but *is*
+      # included in a gemspec, it's because the Gemfile didn't import
+      # the gemspec. This is unusual, but the correct behaviour if/when
+      # it happens is to behave as if the repo was gemspec-only.
+      if dep.nil? && dependency_requirements.any?
+        return "latest"
+      end
+
+      # Otherwise, if the dependency wasn't found it's because it is a
+      # subdependency that was removed when attempting to update it.
+      return nil if dep.nil?
+
+      # If the dependency is Bundler itself then we can't trust the
+      # version that has been returned (it's the version Dependabot is
+      # running on, rather than the true latest resolvable version).
+      return nil if dep.name == "bundler"
+
+      # If the old Gemfile index was used then it won't have checked
+      # Ruby compatibility. Fix that by doing the check manually (and
+      # saying no update is possible if the Ruby version is a mismatch)
+      return nil if ruby_version_incompatible?(dep)
+
+      details = { version: dep.version }
+      if dep.source.instance_of?(::Bundler::Source::Git)
+        details[:commit_sha] = dep.source.revision
+      end
+      details
+    end
+
+    private
+
+    def using_bundler_2?
+      @using_bundler_2
+    end
+
+    def setup_bundler
```

Comment:
This method can probably move to `Functions::setup_bundler(dir:, credentials:, bundler_2:)` or similar as it's used in ~3 places.